src/mc.rs

```diff
@@ -1,12 +1,1025 @@
+            ($args:expr, $name:ident) => {{
+                extern "C" {
+                    fn $name(
+                        tmp: *mut i16,
+                        src: *const <BitDepth16 as BitDepth>::Pixel,
+                        src_stride: ptrdiff_t,
+                        w: c_int,
+                        h: c_int,
+                        mx: c_int,
+                        my: c_int,
+                        dx: c_int,
+                        dy: c_int,
+                        bitdepth_max: c_int,
+                    );
+                }
+
+                let (tmp, src, src_stride, w, h, mx, my, dx, dy, bitdepth_max) = args;
+                $name(tmp, src, src_stride, w, h, mx, my, dx, dy, bitdepth_max)
+            }};
+            ($args:expr, $asm:expr, unreachable) => {{
+                let _ = args;
+                unreachable!("{asm:?}");
+            }};
+        }
+
+        macro_rules! asm_fn {
+            ($asm:expr, $args:expr, $name:ident) => {{
+                use FnAsmVersion::*;
+                paste! {
+                    match asm {
+                        #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+                        SSE2 => extern_fn!($args, asm, unreachable),
+                        #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+                        SSSE3 => extern_fn!($args, [<dav1d_prep $name _16bpc_ssse3>]),
+                        #[cfg(target_arch = "x86_64")]
+                        AVX2 => extern_fn!($args, [<dav1d_prep $name _16bpc_avx2>]),
+                        #[cfg(target_arch = "x86_64")]
+                        AVX512ICL => extern_fn!($args, asm, unreachable),
+                        #[cfg(any(target_arch = "arm", target_arch = "aarch64"))]
+                        Neon => extern_fn!($args, asm, unreachable),
+                    }
+                }
+            }};
+        }
+
+        use Filter8Tap::*;
+        match filter_2d {
+            Filter2dRust::Tap8(Filter2d8Tap { h, v }) => match (h, v) {
+                (Regular, Regular) => asm_fn!(asm, args, _8tap_scaled_regular),
+                (Regular, Smooth) => asm_fn!(asm, args, _8tap_scaled_regular_smooth),
+                (Regular, Sharp) => asm_fn!(asm, args, _8tap_scaled_regular_sharp),
+                (Smooth, Regular) => asm_fn!(asm, args, _8tap_scaled_smooth_regular),
+                (Smooth, Smooth) => asm_fn!(asm, args, _8tap_scaled_smooth),
+                (Smooth, Sharp) => asm_fn!(asm, args, _8tap_scaled_smooth_sharp),
+                (Sharp, Regular) => asm_fn!(asm, args, _8tap_scaled_sharp_regular),
+                (Sharp, Smooth) => asm_fn!(asm, args, _8tap_scaled_sharp_smooth),
+                (Sharp, Sharp) => asm_fn!(asm, args, _8tap_scaled_sharp),
+            },
+            Filter2dRust::BiLinear => asm_fn!(asm, args, _bilin_scaled),
+        }
+    }
+}
+
+#[derive(Copy, Clone, Default)]
+#[repr(C)]
+pub struct Dav1dMCDSPContextRust {
```

Comment:
Can't this be `Dav1dMCDSPContext`? if that name is already taken, can't you get rid of the old one?

Reply:
The old one is still duplicated since it's function pointers haven't been devirtualized yet.  I didn't want to inline each field of the Rust one into all the duplicated old ones, so I put them into one struct, the Rust one.  If there's a better name I could switch to that, but I think it's fine since it's temporary and will go away pretty soon once I finish devirtualizing the rest of the struct.  I can definitely add a comment though.